fix(dropdown): disable host button of disabled dropdown item

jnizet opened the pull request:
Note that a host binding would work fine on buttons but wouldn't on links,
which don't have a disabled property.

fix #4301

Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) and [DEVELOPER.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/DEVELOPER.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [ ] added/updated any applicable API documentation.
 - [ ] added/updated any applicable demos.

jnizet committed:
fix(dropdown): disable host button of disabled dropdown item

Note that a host binding would work fine on buttons but wouldn't on links,
which don't have a disabled property.

fix #4301

maxokorokov reviewed (commented):
Hey, @jnizet!

Thanks for the PR, I was thinking about this a bit yesterday. The thing is that `ngbDropdownItem` is not limited to buttons, but works on both `<button>` and `<a>` elements. As well as bootstrap styles, and moreover they use mostly `<a>`s in their documentation.

So I have a couple of questions here:
1. should we add `disabled` property on a button → I'd say yes, it is confusing now
2. should we add `disabled` property on a link → probably no, but no harm done I guess?
4. should we "disable" a link with say `tablindex` → not sure

This LGTM for buttons, but I'm not sure about link. WDYT?

jnizet commented:
Hi @maxokorokov 

My first idea was to test whether the host element was a button before setting the property. But I wasn't sure if that would work fine with SSR. So I tried just setting the property without testing the type of the host element. And apparently, `Renderer2.setProperty()` doesn't do anything if the property doesn't exist. BTW, the PR adds a test with buttons and another test with links, and both pass without emitting any error or warning.

I also tried with a host binding BTW, and that fails with links (but works with buttons).

Regarding tabindex, that's a good point: I don't think anyone will ever want to customize the tab index on dropdown items, so setting it to -1 when disabled and unsetting it when enabled is probably a good idea. 

WDYT?

maxokorokov commented:
@jnizet, thanks for explanations, didn't know about the renderer.

The more I think about it, the more I confused I get :)

A couple more things I've learned:
- bootstrap emphasise in docs they only make an item "appear" disabled with `.disabled` (visually disabled). And don't care about anything else like functionally or semantically disabling items.
- there is `aria-disabled="true"` we forgot about (semantically disabled)
- to disable (functionally) a link the common recommentdation is to remove the `href` completely, but I don't think we want to mess with `href` (ex. can have routerlinks, can be dynamic, etc)
- bootstrap has confusing keyboard navigation (arrowup/down between non-disabled dropdown items vs Tabs) :(

It seems to me we have two options:

1. Truly disabling an item (straightforward one)
```html
<!-- adding 'disabled' to the button, adding tabindex too for simplicity won't hurt -->
<button class="dropdown-item disabled" disabled>

<!-- removing href from the link is better, but this would do I guess -->
<a href="#" class="dropdown-item disabled" tabindex="-1">
```

bad: not accessible at all for screen readers, neither with Tab nor with arrows

2. Marking an item as disabled and documenting the rest (a11y-focused)

```html
<!-- just adding aria-label and leaving  -->
<button class="dropdown-item disabled" aria-disabled="true">
<a href="#" class="dropdown-item disabled" aria-disabled="true">
```

bad: not fully functionally disabled, can select with keyboard still; confusion with `<button [disabled]="true">` not adding the disabled property.

---

I think at the moment I'd prefer the option 1 and just adding the `tabindex="-1"` in the PR.

---

Just one use-case that won't work is that if somebody wants to have a11y AND arrowup/down keyboard navigation:

```html
<!-- This won't work, because arrowup/down works based on 
@ContentChildren(NgbDropdownItem) that have `disabled` field === false  -->
<a href="#" role="button" ngbDropdownItem [class.disabled]="disabled" [attr.aria-disabled]="disabled">
```
https://github.com/ng-bootstrap/ng-bootstrap/blob/108fa52e727f8539b9b82a20d51d6193b00119cc/src/dropdown/dropdown.ts#L444

jnizet commented:
@maxokorokov I also prefer option 1. 

But why not combine it with option 2 and also add aria-disabled="true" on the button or link when it's disabled? TBH, I'm still baffled that using disabled binding binds the input of the directive but not the property of the button. I guess it's a feature and not a bug of Angular, but it's very surprising.

maxokorokov commented:
I think `aria-disabled` won't do much, if you can't focus the element?

jnizet commented:
Alright, I see. I will add a commit which also sets tabindex.

jnizet committed:
fix(dropdown): set tabIndex to -1 on disabled dropdown items

This makes sure that disabled dropdown items using links are not tab-traversable.

fix #4301

maxokorokov commented on src/dropdown/dropdown.ts:
```diff
@@ -49,11 +52,15 @@ export class NgbDropdownItem {
   @Input()
   set disabled(value: boolean) {
     this._disabled = <any>value === '' || value === true;  // accept an empty attribute as true
+    // note: we don't use a host binding for disabled because when used on links, it fails because links don't have a
```
👍

maxokorokov reviewed (approved):
LGTM!
